

rule download_fasta:
    output:
        '{directory}/annotations/fa.fa.gz'
    params:
        species = config['genome']['species']
    log: '{directory}/annotations/species.txt'
    shell:
        'wget https://hgdownload.soe.ucsc.edu/goldenPath/{params.species}/bigZips/{params.species}.fa.gz '
        '-O {output} && echo {params.species} > {log}'


def get_fasta_file(wildcards):

    if 'fasta' in config['genome'] and not config['genome']['fasta'] is None:
        return config['genome']['fasta']
    elif 'species' in config['genome']:
        return rules.download_fasta.output
    else:
        raise ValueError('Neither "fasta" nor "species" configured. Cannot obtain fasta file.')


rule get_chromsizes:
    input:
        get_fasta_file
    output:
        '{directory}/annotations/chrom_sizes.txt'
    shell:
        'faSize -detailed -tab {input} > {output}'


rule download_gtf:
    output:
       '{directory}/annotations/refGene.gtf.gz'
    params:
        species = config['genome']['species']
    shell:
        'wget https://hgdownload.soe.ucsc.edu/goldenPath/{params.species}/bigZips/genes/refGene.gtf.gz '
        '-O {output}'